File: C++/findMinHeightTrees.cpp
/**
 * @File    : findMinHeightTrees.cpp
 * @Brief   : 最小高度树
 * @Link    : https://leetcode-cn.com/problems/minimum-height-trees/
 * @Author  : Wei Li
 * @Date    : 2021-07-12
*/

#include <iostream>
#include <vector>
#include <queue>

// 拓扑排序
class Solution
{
public:
    std::vector<int> findMinHeightTrees(int n, std::vector<std::vector<int>> &edges)
    {
        if (n == 1)
            return {0};
        // 根据边表edges 创建图(邻接数组)
        std::vector<std::vector<short>> adjList(n);
        std::vector<short> degree(n, 0);
        for (short i = 0; i < edges.size(); ++i)
        {
            short v1 = edges[i][0];
            short v2 = edges[i][1];
            adjList[v1].push_back(v2);
            adjList[v2].push_back(v1);

            ++degree[v1];
            ++degree[v2];
        }

        // BFS拓扑排序
        // 初始化
        std::queue<short> Q;
        for (short i = 0; i < n; ++i)
        {
            if (degree[i] == 1)
            {
                Q.push(i);
            }
        }

        while (Q.size())
        {
            short size = Q.size();
            if (size == n)
                break; // 终止条件
            n -= size;

            while (size--)
            {
                short cur = Q.front();
                Q.pop();

                // 访问邻接点，找出正确的下一个adj
                short adj;
                for (short i = 0; i < adjList[cur].size(); ++i)
                {
                    adj = adjList[cur][i];
                    if (degree[adj] > 1)
                        break; // 度=1的是已经被访问过的
                }
                --degree[adj];

                if (degree[adj] == 1)
                {
                    Q.push(adj);
                }
            }
        }

        std::vector<int> ans;
        while (Q.size())
        {
            ans.push_back(Q.front());
            Q.pop();
        }

        return ans;
    }
};

// ------------------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}

File: Python/findMinHeightTrees.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findMinHeightTrees.py
@Function: 最小高度树
@Link: https://leetcode-cn.com/problems/minimum-height-trees/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-12
"""

import collections
from typing import List


class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        # 简单无向图：套路是建图并遍历
        # 建图：邻接表
        # 邻接表为map,其值为list,它的size就是入度数
        if n == 2:
            return [0, 1]
        if n == 1:
            return [0]

        adjs = collections.defaultdict(list)  # defaultdict写法很有用
        for x in edges:  # 图的邻接表表示法,基本是模板
            adjs[x[0]].append(x[1])  # 1:{2}
            adjs[x[1]].append(x[0])  # 2:{1}

        # BFS队列: 初始队列放入初始元素,size=1的为叶子,入队
        queue = collections.deque()  # 固定写法
        for key, value in adjs.items():
            if len(value) == 1:
                queue.append(key)

        # BFS两个大循环
        while(queue):  # 固定写法
            size = len(queue)  # 固定写法
            n = n - size

            for _ in range(size):
                v = queue.popleft()
                v_adj = adjs[v].pop()  # v的邻接仅一个,弹出即删除
                adjs[v_adj].remove(v)  # 在v的邻接元素的邻接列表里删除v
                if len(adjs[v_adj]) == 1:
                    queue.append(v_adj)

            if n == 1:
                return [queue.popleft()]
            if n == 2:
                return [queue.popleft(), queue.popleft()]


# --------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass
